File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/share/AddPresenter.java
package app.editors.manager.mvp.presenters.share;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import app.editors.manager.R;
import app.editors.manager.app.App;
import app.editors.manager.managers.utils.CollectionUtils;
import app.editors.manager.mvp.models.base.Entity;
import app.editors.manager.mvp.models.explorer.File;
import app.editors.manager.mvp.models.explorer.Folder;
import app.editors.manager.mvp.models.explorer.Item;
import app.editors.manager.mvp.models.list.Header;
import app.editors.manager.mvp.models.models.ModelShareStack;
import app.editors.manager.mvp.models.request.RequestShare;
import app.editors.manager.mvp.models.request.RequestShareItem;
import app.editors.manager.mvp.models.response.ResponseGroups;
import app.editors.manager.mvp.models.response.ResponseShare;
import app.editors.manager.mvp.models.response.ResponseUsers;
import app.editors.manager.mvp.models.user.Group;
import app.editors.manager.mvp.models.user.User;
import app.editors.manager.mvp.presenters.base.BasePresenter;
import app.editors.manager.mvp.views.share.AddView;
import app.editors.manager.ui.fragments.share.AddFragment;
import moxy.InjectViewState;
import retrofit2.Call;
import retrofit2.Response;

@InjectViewState
public class AddPresenter extends BasePresenter<AddView, ResponseUsers> {

    public static final String TAG = AddPresenter.class.getSimpleName();

    private final String mToken;
    private Item mItem;
    private AddFragment.Type mType;
    private ModelShareStack mModelShareStack;
    private boolean mIsCommon;
    private int mCommonResponse;
    private String mSearchValue;

    private Call<ResponseUsers> mResponseUsersCall;
    private Call<ResponseGroups> mResponseGroupsCall;
    private Call<ResponseShare> mResponseAddCall;

    public AddPresenter() {
        App.getApp().getAppComponent().inject(this);
        mModelShareStack = ModelShareStack.getInstance();
        mToken = mPreferenceTool.getToken();
        mIsCommon = false;
        mCommonResponse = 0;
    }

    /*
    * Requests
    * */
    private void getUsers() {
        mIsCommon = false;
        mResponseUsersCall = mRetrofitTool.getApiWithPreferences().getUsers(mToken, getArgs());
        mResponseUsersCall.enqueue(new CallbackUsersCallback());
    }

    private void getGroups() {
        mIsCommon = false;
        mResponseGroupsCall = mRetrofitTool.getApiWithPreferences().getGroups(mToken, getArgs());
        mResponseGroupsCall.enqueue(new CallbackGroupsResponse());
    }

    public void getCommons() {
        mIsCommon = true;
        mCommonResponse = 0;
        mResponseUsersCall = mRetrofitTool.getApiWithPreferences().getUsers(mToken, getArgs());
        mResponseUsersCall.enqueue(new CallbackUsersCallback());
        mResponseGroupsCall = mRetrofitTool.getApiWithPreferences().getGroups(mToken, getArgs());
        mResponseGroupsCall.enqueue(new CallbackGroupsResponse());
    }

    private void shareFileTo(final String id) {
        final RequestShare requestShare = getRequestShare();
        if (requestShare != null) {
            mResponseAddCall = mRetrofitTool.getApiWithPreferences().setFileAccess(mToken, id, requestShare);
            mResponseAddCall.enqueue(new CallbackAddResponse());
        }
    }

    private void shareFolderTo(final String id) {
        final RequestShare requestShare = getRequestShare();
        if (requestShare != null) {
            mResponseAddCall = mRetrofitTool.getApiWithPreferences().setFolderAccess(mToken, id, requestShare);
            mResponseAddCall.enqueue(new CallbackAddResponse());
        }
    }

    @Nullable
    private RequestShare getRequestShare() {
        final List<RequestShareItem> shareItems = new ArrayList<>();

        // Get users access list
        for (User item : mModelShareStack.getUserSet()) {
            if (item.isSelected()) {
                final String idItem = item.getId();
                final int accessCode = mModelShareStack.getAccessCode();
                final RequestShareItem requestShareItem = new RequestShareItem();
                requestShareItem.setShareTo(idItem);
                requestShareItem.setAccess(String.valueOf(accessCode));
                shareItems.add(requestShareItem);
            }
        }

        // Get groups access list
        for (Group item : mModelShareStack.getGroupSet()) {
            if (item.isSelected()) {
                final String idItem = item.getId();
                final int accessCode = mModelShareStack.getAccessCode();
                final RequestShareItem requestShareItem = new RequestShareItem();
                requestShareItem.setShareTo(idItem);
                requestShareItem.setAccess(String.valueOf(accessCode));
                shareItems.add(requestShareItem);
            }
        }

        if (!shareItems.isEmpty()) {
            final RequestShare requestShare = new RequestShare();
            requestShare.setShare(shareItems);
            final String message = mModelShareStack.getMessage();
            if (message != null && !message.isEmpty()) {
                requestShare.setNotify(true);
                requestShare.setSharingMessage(message);
            }

            return requestShare;
        }

        return null;
    }

    private List<Entity> getCommonList() {
        final List<Entity> commonList = new ArrayList<>();

        // Users
        if (!mModelShareStack.isUserEmpty()) {
            commonList.add(new Header(mContext.getString(R.string.share_add_common_header_users)));
            commonList.addAll(getUserListItems());
        }

        // Groups
        if (!mModelShareStack.isGroupEmpty()) {
            commonList.add(new Header(mContext.getString(R.string.share_add_common_header_groups)));
            commonList.addAll(getGroupListItems());
        }

        return commonList;
    }

    public void getShared() {
        switch (mType) {
            case USERS:
                getUsers();
                break;
            case GROUPS:
                getGroups();
                break;
        }
    }

    public void shareItem() {
        if (mItem instanceof Folder) {
            shareFolderTo(mItem.getId());
        } else if (mItem instanceof File) {
            shareFileTo(mItem.getId());
        }
    }

    /*
    * Update states
    * */
    public void updateTypeSharedListState() {
        switch (mType) {
            case USERS:
                getViewState().onGetUsers(getUserListItems());
                break;
            case GROUPS:
                getViewState().onGetGroups(getGroupListItems());
                break;
        }
    }

    public void updateCommonSharedListState() {
        getViewState().onGetCommon(getCommonList());
    }

    public void updateSearchState() {
        getViewState().onSearchValue(mSearchValue);
    }

    /*
    * Getters/Setters
    * */
    public void setItem(final Item item) {
        mItem = item;
    }

    public void setType(final AddFragment.Type type) {
        mType = type;
    }

    private Map<String, String> getArgs() {
        final Map<String, String> args = new TreeMap<>();
        return args;
    }

    private List<Entity> getUserListItems() {
        return CollectionUtils.convertUsersToItems(mModelShareStack.getUserSet());
    }

    private List<Entity> getGroupListItems() {
        return CollectionUtils.convertGroupsToItems(mModelShareStack.getGroupSet());
    }

    public int getCountChecked() {
        return mModelShareStack.getCountChecked();
    }

    public void resetChecked() {
        mModelShareStack.resetChecked();
    }

    public void setAccessCode(final int accessCode) {
        mModelShareStack.setAccessCode(accessCode);
    }

    public int getAccessCode() {
        return mModelShareStack.getAccessCode();
    }

    public void setMessage(final String message) {
        mModelShareStack.setMessage(message);
    }

    @Nullable
    public void setSearchValue(String searchValue) {
        mSearchValue = searchValue;
    }

    /*
    * Common callback for users
    * */
    private class CallbackUsersCallback extends CommonCallback<ResponseUsers> {

        @Override
        public void onSuccessResponse(Response<ResponseUsers> response) {
            if (!mResponseUsersCall.isCanceled()) {
                mModelShareStack.addUsers(response.body().getResponse());
                mModelShareStack.removeById(mPreferenceTool.getSelfId());
                if (mIsCommon) {
                    if (++mCommonResponse == 2) {
                        getViewState().onGetCommon(getCommonList());
                    }
                } else {
                    getViewState().onGetUsers(getUserListItems());
                }
            }
        }
    }

    /*
    * Common callback for groups
    * */
    private class CallbackGroupsResponse extends CommonCallback<ResponseGroups> {

        @Override
        public void onSuccessResponse(Response<ResponseGroups> response) {
            if (!mResponseGroupsCall.isCanceled()) {
                mModelShareStack.addGroups(response.body().getResponse());
                if (mIsCommon) {
                    if (++mCommonResponse == 2) {
                        getViewState().onGetCommon(getCommonList());
                    }
                } else {
                    getViewState().onGetGroups(getGroupListItems());
                }
            }
        }
    }

    /*
    * Common on added
    * */
    private class CallbackAddResponse extends CommonCallback<ResponseShare> {

        @Override
        public void onSuccessResponse(Response<ResponseShare> response) {
            if (!mResponseAddCall.isCanceled()) {
                mModelShareStack.resetChecked();
                getViewState().onSuccessAdd();
            }
        }
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/share/AddPresenter.kt
package app.editors.manager.mvp.presenters.share

import app.documents.core.network.common.contracts.ApiContract
import app.documents.core.network.common.extensions.checkStatusCode
import app.documents.core.network.common.extensions.request
import app.documents.core.network.manager.models.explorer.CloudFile
import app.documents.core.network.manager.models.explorer.CloudFolder
import app.documents.core.network.manager.models.explorer.Item
import app.documents.core.network.share.ShareService
import app.documents.core.network.share.models.request.RequestShare
import app.documents.core.network.share.models.request.RequestShareItem
import app.documents.core.storage.account.CloudAccount
import app.editors.manager.R
import app.editors.manager.app.App
import app.editors.manager.app.appComponent
import app.editors.manager.app.roomProvider
import app.editors.manager.app.shareApi
import app.editors.manager.mvp.models.models.ModelShareStack
import app.editors.manager.mvp.models.ui.GroupUi
import app.editors.manager.mvp.models.ui.ShareHeaderUi
import app.editors.manager.mvp.models.ui.UserUi
import app.editors.manager.mvp.presenters.base.BasePresenter
import app.editors.manager.mvp.views.share.AddView
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import lib.toolkit.base.ui.adapters.holder.ViewType
import moxy.InjectViewState
import moxy.presenterScope
import java.io.Serializable

@InjectViewState
class AddPresenter(
    val item: Item,
    val type: Type
) : BasePresenter<AddView>() {

    sealed class Type : Serializable {
        data object Users : Type()
        data object Groups : Type()
        data object Common : Type()
    }

    companion object {
        val TAG: String = AddPresenter::class.java.simpleName
    }

    private val shareStack: ModelShareStack = ModelShareStack.getInstance()
    private var searchValue: String? = null
    private var job: Job? = null

    private val userListItems: List<UserUi>
        get() = shareStack.userSet.sortedBy { it.displayName }

    private val groupListItems: List<GroupUi>
        get() = shareStack.groupSet.sortedWith(groupComparator())

    val countChecked: Int
        get() = shareStack.countChecked

    val isSelectedAll: Boolean
        get() = when (type) {
            Type.Users -> shareStack.userSet.any { !it.isSelected }
            Type.Groups -> shareStack.groupSet.any { !it.isSelected }
            else -> false
        }

    val isSelected: Boolean
        get() = when (type) {
            Type.Users -> shareStack.userSet.any { it.isSelected }
            Type.Groups -> shareStack.groupSet.any { it.isSelected }
            else -> false
        }

    var accessCode: Int
        get() = shareStack.accessCode
        set(accessCode) {
            shareStack.accessCode = accessCode
        }

    init {
        App.getApp().appComponent.inject(this)
    }

    private val account: CloudAccount =
        context.appComponent.accountOnline ?: throw RuntimeException("Account can't be null")

    private val shareApi: ShareService = context.shareApi

    override fun onDestroy() {
        super.onDestroy()
        job = null
    }

    private suspend fun getUsers(searchValue: String = ""): List<UserUi> {

        val invitedUsersId = if ((item as? CloudFolder)?.isRoom == true) {
            context.roomProvider.getRoomUsers(item.id).map { it.sharedTo.id }
        } else {
            emptyList()
        }

        return shareApi.getUsers(getOptions(searchValue))
            .checkStatusCode(::fetchError)
            .response
            .filter { user ->
                user.id != account.id
                        && user.id != item.createdBy.id
                        && user.displayName.isNotEmpty()
                        && !invitedUsersId.contains(user.id)
                        && user.activationStatus == 1
            }
            .map { user ->
                UserUi(
                    id = user.id,
                    department = user.department,
                    displayName = user.displayName,
                    avatarUrl = user.avatar,
                    status = user.activationStatus
                )
            }.sortedBy { it.status }
    }

    private suspend fun getGroups(searchValue: String = ""): List<GroupUi> {
        return shareApi.getGroups(getOptions(searchValue, true))
            .checkStatusCode(::fetchError)
            .response
            .map {
                GroupUi(
                    id = it.id,
                    name = it.name,
                    manager = it.manager.orEmpty()
                )
            }
            .toMutableList()
            .also { list ->
                list.add(GroupUi(GroupUi.GROUP_ADMIN_ID, "", ""))
                list.add(GroupUi(GroupUi.GROUP_EVERYONE_ID, "", ""))
            }

    }

    private suspend fun shareFileTo(id: String) {
        requestShare?.let { request ->
            request(
                func = { shareApi.setFileAccess(id, request) },
                onSuccess = {
                    shareStack.resetChecked()
                    viewState.onSuccessAdd()
                },
                onError = ::fetchError
            )
        }
    }

    private suspend fun shareFolderTo(id: String) {
        requestShare?.let { request ->
            request(
                func = { shareApi.setFolderAccess(id, request) },
                onSuccess = {
                    shareStack.resetChecked()
                    viewState.onSuccessAdd()
                }, onError = ::fetchError
            )
        }
    }

    private val requestShare: RequestShare?
        get() {
            val shareItems: MutableList<RequestShareItem> = ArrayList()

            // Get users access list
            for (item in shareStack.userSet) {
                if (item.isSelected) {
                    val idItem = item.id
                    val accessCode = shareStack.accessCode
                    shareItems.add(RequestShareItem(shareTo = idItem, access = accessCode.toString()))
                }
            }

            // Get groups access list
            for (item in shareStack.groupSet) {
                if (item.isSelected) {
                    val idItem = item.id
                    val accessCode = shareStack.accessCode
                    shareItems.add(RequestShareItem(shareTo = idItem, access = accessCode.toString()))
                }
            }
            if (shareItems.isNotEmpty()) {
                val message = shareStack.message
                return RequestShare(
                    share = shareItems,
                    isNotify = !message.isNullOrEmpty(),
                    sharingMessage = message ?: ""
                )
            }
            return null
        }

    private val commonList: List<ViewType>
        get() {
            val commonList: MutableList<ViewType> = ArrayList()

            // Users
            if (!shareStack.isUserEmpty) {
                commonList.add(ShareHeaderUi(context.getString(R.string.share_add_common_header_users)))
                commonList.addAll(userListItems)
            }

            // Groups
            if (!shareStack.isGroupEmpty) {
                commonList.add(ShareHeaderUi(context.getString(R.string.share_add_common_header_groups)))
                commonList.addAll(groupListItems)
            }
            return commonList
        }

    private fun getOptions(value: String, isGroup: Boolean = false): Map<String, String> =
        mapOf(
            ApiContract.Parameters.ARG_FILTER_VALUE to value,
            ApiContract.Parameters.ARG_FILTER_BY to if (isGroup) ApiContract.Parameters.VAL_SORT_BY_NAME else ApiContract.Parameters.VAL_SORT_BY_DISPLAY_NAME,
            ApiContract.Parameters.ARG_FILTER_OP to ApiContract.Parameters.VAL_FILTER_OP_CONTAINS
        )

    private fun groupComparator() = Comparator<GroupUi> { a, b ->
        val list = listOf(GroupUi.GROUP_ADMIN_ID, GroupUi.GROUP_EVERYONE_ID)
        when {
            a.id in list -> -1
            b.id in list -> 1
            else -> a.name.lowercase().compareTo(b.name.lowercase())
        }
    }

    fun fetchSharedList(searchValue: String = "") {
        presenterScope.launch {
            when (type) {
                Type.Common -> {
                    shareStack.clearModel()
                    shareStack.addGroups(getGroups(searchValue))
                    shareStack.addUsers(getUsers(searchValue))
                    viewState.onGetCommon(commonList)
                }
                Type.Groups -> {
                    shareStack.addGroups(getGroups(searchValue))
                    viewState.onGetGroups(groupListItems)
                }
                Type.Users -> {
                    shareStack.clearModel()
                    shareStack.addUsers(getUsers(searchValue))
                    viewState.onGetUsers(userListItems)
                }
            }
        }
    }

    fun shareItem() {
        presenterScope.launch {
            when (item) {
                is CloudFolder -> shareFolderTo(item.id)
                is CloudFile -> shareFileTo(item.id)
            }
        }
    }

    fun updateTypeSharedListState() {
        when (type) {
            Type.Common -> viewState.onGetCommon(commonList)
            Type.Groups -> viewState.onGetGroups(groupListItems)
            Type.Users -> viewState.onGetUsers(userListItems)
        }
    }

    fun updateSearchState() {
        viewState.onSearchValue(searchValue)
    }

    fun resetChecked() {
        shareStack.resetChecked()
    }

    fun setMessage(message: String?) {
        shareStack.message = message
    }

    fun setSearchValue(searchValue: String?) {
        this.searchValue = searchValue
        job?.cancel()
        job = presenterScope.launch {
            delay(350L)
            fetchSharedList(searchValue.orEmpty())
        }
    }
}